Translation - scripts, the final part

marmarek opened the pull request:
This is the final part of the translation integration. Previous two:
https://github.com/QubesOS/qubesos.github.io/pull/153
https://github.com/QubesOS/qubes-doc/pull/1123

Similar to previous parts, the majority of work was done by @maiska and @tokideveloper. Thanks!

On top of their work, this PR integrates all those scripts with Gitlab CI pipeline. It handles pushing/pulling content to/from Transifex automatically. I have performed the first iteration of this, to populate https://github.com/QubesOS/qubes-translated repository with initial content (we did have some content translated before, so indeed some (small) parts are translated, mostly in Spanish language). For the test purposes, I've enabled languages `es`, `de`, `fr`, but this commit should not be merged until substantial part of the documentation there is translated (right now most of those pages are simply copies from English).

See README for high level description of how it works.

The resulting website, with translations enabled can be seen at https://wwwtest.qubes-os.org/

Commit:
Add translation scripts

Co-authored-by: Tobias Killer <tokidev@posteo.de>

marmarek committed:
Add qubes-translated submodule

Add it as "_translated" for consistency with other submodules. This is a
change from _qubes-translated name in the scripts, so update the scripts
(and few html files) here too.

Adjust tx_confg.sh script for few renames.

marmarek committed:
Various translation scripts improvements

- use `set -e` in top level script, quote variables
- prefer env variable for API token (not visible in the process list,
  and easier to provide in CI config); same for TRANSLATED_LANGS setting
- fix setting <a id=...> tags - there need to be an empty line before
  the content, otherwise the header is not handled correctly; this
  change makes merge_md_heading_ids.rb not idempotent, because line
  numbers change now, but it should be fine to call it one time directly
  after tx pull
- if htmlproofer says it's ok, don't try to recover from (non-existent) failure
- fix setting ref/lang (erroneously commented out before)
- handle redirect_to frontmatter key (if internal - prefix with lang
  too)
- better synchronize translated frontmatter to the source one - if
  source lacked redirect_from, remove one from translated too
- use `tx pull --force` - git checkout time isn't necessary time of the
  last `tx pull`, so comparing mtime isn't really helpful to decide
  which file is up to date
- don't prepend language prefix when redirecting to news article - those
  are not translated
- sort redirect_from frontmatter entry, to avoid no-op commits

marmarek committed:
Introduce an option to filter what languages are visible in the switcher

Also, if there is just one, hide the language switcher

marmarek committed:
Update translated files counter

marmarek committed:
Add initial symlinks in _data/translation

For now symlink es, de, fr. We can add more when we'll have
translations.

marmarek committed:
Add ci job for pushing/pulling content to Transifex

Add two wrapper scripts in _utils for that.

marmarek committed:
Reduce verbosity - CI log is too long

marmarek committed:
Add scripts to fetch+update qubes-translated submodule

Fetch the update made by _utils/transifex-* scripts. The transifex-pull
script performs quite detailed inspections and apply fixups. At this
stage apply only a basic sanity check with the sole purpose: do not
allow translated content to subvert origin english one.

marmarek committed:
autoupdate: _translated

marmarek committed:
[DO NOT MERGE] enable languages: de fr es

marmarek committed:
Add info on Transifex integration to the README

marmarek commented:
testdeploy

tokideveloper committed:
Add `Translate This Page` button

(rebased onto newer doc-widget.html by @marmarek)

marmarek committed:
Skip "Edit This Page" button on translated pages

qubes-translated repo should not be modified directly.

tokideveloper commented:
Great work! Thanks to @maiska and @marmarek!

BTW, the [line](https://github.com/QubesOS/qubesos.github.io/pull/154/commits/481cc696cc32131fbd1f52fc5c0427a7cbb604c5#diff-c0311c1df17d5d714db46ca1401ad374a35de861a461d211c6c947f1b2194034R11)
```
{% assign langmenu = (site.languages.size > 1) %}
```

doesn't seem to work as expected. At least, I see the lang menu with the English entry only. The following should work (without having tested it):

```
{% if site.languages.size > 1 %}
  {% assign langmenu = true %}
{% else %}
  {% assign langmenu = false %}
{% endif %}
```

It's a Liquid thing …

marmarek commented:
> `{% assign langmenu = (site.languages.size > 1) %}`

It works, just ignore wwwpreview instance - it got built with empty _translated dir... See wwwtest instance, linked in the PR description.

tokideveloper commented:
> > `{% assign langmenu = (site.languages.size > 1) %}`
> 
> It works, just ignore wwwpreview instance - it got built with empty _translated dir... See wwwtest instance, linked in the PR description.

I see multiple langs in the menu in the `wwwtest` instance, so, it seems to work as you said.

However, I tried the following mini examples right now:

**Example 1**
```
{% assign a = (10 > 5) %}
{{ a }}
```
outputs *nothing*. Also, I got a syntax error warning.

**Example 2**
```
{% assign a = 10 > 5 %}
{{ a }}
```
outputs `10`. Also, I got a syntax error warning.

**Example 3**
```
{% if 10 > 5 %}
  {% assign a = true %}
{% else %}
  {% assign a = false %}
{% endif %}
{{ a }}
```
outputs `true`. No warning.

**Example 4**
```
{% if (10 > 5) %}
  {% assign a = true %}
{% else %}
  {% assign a = false %}
{% endif %}
{{ a }}
```
outputs `false`. Also, I got a syntax error warning.

Do I have a different Liquid version?
What's the output of `{{ langmenu }}`?

Sorry for the noise. I just want to avoid errors in the long run.

marmarek commented:
You are correct, it worked only because `langmenu` was assigned `4`. Thanks!

tokideveloper commented on _utils/_translation_utils/post_transifex_pull.sh:
```diff
@@ -0,0 +1,48 @@
+
+echo "============================ post processing step 2 ======================================"
+#read b
+ruby _utils/_translation_utils/merge_md_heading_ids.rb "$1" /tmp/tx-mapping
```
At least for the `merge_md_heading_ids.rb` script, called here, it is important that the checked out commit is not the most recent one but that commit that was the current one when the files have been pushed to Transifex the last time. The reason is that headlines may have changed or added or removed meanwhile and thus, the `merge_md_heading_ids.rb` script may produce a wrong output.

I can't see any attempt to checkout that correct commit or have I overseen something?

IMHO when pushing files to Transifex, the related commit hash should be saved. When pulling the files from Transifex, that related (old) commit should be checked out for the postprocessing steps.

Is it understandable what I mean?

marmarek replied:
This all is called just after _pushing_ and pulling from transifex. See https://github.com/QubesOS/qubesos.github.io/pull/154/files#diff-037ea159eb0a7cb0ac23b851e66bee30fb838ee8d0d99fa331a1ba65283d37f7R38-R39

tokideveloper replied:
Ah, okay, now I see. Both pushing and pulling belong to the same commit and thus everything is right. Thank you for clarification!

tokideveloper committed:
Update merge_md_heading_ids.rb script

marmarek committed:
Make Q logo link to the main page in the current language

marmarek committed:
translation: remove old translations

Remove translated files when the source file get removed. This includes
also renaming files.

marmarek commented:
I've rebased it on top of recent version and added few improvement discussed in the meantime (including @tokideveloper 's updated script).
This PR includes also https://github.com/QubesOS/qubesos.github.io/pull/177

andrewdavidwong commented on _includes/doc-widget.html:
```diff
@@ -25,7 +30,13 @@
   {% if item.icon == "fa-code" %}
     {% assign a_href = master_blob %}
   {% elsif item.icon == "fa-code-fork" %}
+    {% if lang != "" %}
+      <!-- skip "Edit This Page" button on translated pages, link to Transifex only -->
+      {% continue %}
+    {% endif %}
```
But our English doc pages have `lang: en`. Won't this skip the "Edit This Page" button on the English pages too?

marmarek replied:
No, `lang` is a variable with URL prefix, set to empty if `page.lang` is either empty or `en`.

andrewdavidwong commented on _includes/header.html:
```diff
@@ -25,18 +29,25 @@
       {% if langmenu %}
-        {% assign posts = site.pages | concat: site.doc | concat: site.qubes-translated | where:'ref', page.ref | sort: 'lang' %}
+        {% assign posts = site.pages | concat: site.doc | concat: site.translated | where:'ref', page.ref | sort: 'lang' %}
         <div class="langdropdown langdropdown-sm">
           <label class="langdropbtn-sm" for="langdropdown-toggle-sm"><img src="/attachment/icons/icons8-translation-64.png" width="24" height="24"></label>
           <input type="checkbox" id="langdropdown-toggle-sm">
           <div class="langdropdown-content" id="langdropdown-dd-sm">
             {% for post in posts %}
-              <a class="c1" href="{{ post.url }}">{{ post.lang }}</a>
+              {% if site.languages contains post.lang %}
+                <a class="c1" href="{{ post.url }}">{{ post.lang }}</a>
+              {% endif %}
```
I thought we weren't doing translations for posts?

marmarek replied:
No, it isn't really about news, but about any page. Variable name is confusing here indeed. `posts` is in fact all the pages (docs, and base website pages).

andrewdavidwong commented on _includes/news-post-author.html:
```diff
@@ -5,7 +5,7 @@
+{% assign t = site.pages | concat: site.doc | concat: site.translated | where:'permalink', team_link %}
 {% if t.size == 0 %}
   {% assign team_link = "/team/" %}
 {% endif %}
```
What's the purpose of this? Does it just translate `/team/`?

andrewdavidwong replied:
Never mind. See https://github.com/QubesOS/qubesos.github.io/pull/178.

andrewdavidwong commented on _includes/news.html:
```diff
@@ -11,7 +11,7 @@
 {% endif %}
 {% assign news = inc | where_exp: "item", "item.section == 'news'" | first  %}
 {% assign team_link = lang | append: "/team/" %}
-{% assign t = site.pages | concat: site.qubes-translated | where:'permalink', team_link %}
+{% assign t = site.pages | concat: site.translated | where:'permalink', team_link %}
 {% if t.size == 0 %}
   {% assign team_link = "/team/" %}
 {% endif %}
```
This looks like a repetition of the code above.

andrewdavidwong replied:
Addressed by https://github.com/QubesOS/qubesos.github.io/pull/178.

andrewdavidwong commented on _includes/team-heading.html:
```diff
@@ -11,7 +11,7 @@
 {% endif %}
 {% assign upperlinks = teamed | where_exp: "item", "item.htmlsection == 'upperlinks'" | first  %}
 {% assign team_link = lang | append: "/team/" %}
-{% assign teams = site.pages | concat: site.qubes-translated | where:'permalink', team_link %}
+{% assign teams = site.pages | concat: site.translated | where:'permalink', team_link %}
 {% if teams.size == 0 %}
   {% assign team_link = "/team/" %}
 {% endif %}
```
This also looks like (mostly) a repetition of the code above.

andrewdavidwong replied:
Addressed by https://github.com/QubesOS/qubesos.github.io/pull/178.

andrewdavidwong commented on _includes/team-simple.html:
```diff
@@ -6,7 +6,7 @@
   {% assign lang = "/" | append: page.lang %}
 {% endif %}
 {% assign team_link = lang | append: "/team/" %}
-{% assign t = site.pages | concat: site.qubes-translated | where:'permalink', team_link %}
+{% assign t = site.pages | concat: site.translated | where:'permalink', team_link %}
 {% if t.size == 0 %}
   {% assign team_link =  "/team/" %}
 {% endif %}
```
This also looks like (mostly) a repetition of the code above.

andrewdavidwong replied:
Addressed by https://github.com/QubesOS/qubesos.github.io/pull/178.

marmarek commented:
I'll simplify it further when rebasing on top of #178 (when it will be merged).

marmarek committed:
Clarify variable name in lang switcher

marmarek committed:
translation: Move site verification after downloading all languages

Do not attempt to verify correctness of the website before refreshing
all languages. If any structural change happened (rename, permalink
change etc), it will fail because of the old files, not the freshly
downloaded ones.

Make htmlproofer postprocessing script language agnostic. This is mostly
about removing "language" parameter, which wasn't used anyway.

marmarek commented:
PipelineRefresh

marmarek commented:
PipelineRetry

marmarek commented:
PipelineRetry

marmarek commented:
PipelineRetry

andrewdavidwong commented:
Just to confirm... you're not waiting or expecting me to do anything with this, right @marmarek? I've been intentionally leaving this PR alone because you seem like you're managing it yourself (and of course the big "[DO NOT MERGE]" warning).

marmarek commented:
testdeploy

marmarek commented:
trestdeploy

marmarek commented:
testdeploy

marmarek commented:
You are correct. I want to get it working again first. It's almost there. For example https://github.com/QubesOS/qubes-translated/commit/a102a9e3cc6bfe7a1a73c647a18c0406e97c0616 was just made. There is still some issue with language switcher (just "en" visible there).

As for the "[DO NOT MERGE]" commits - it's just about that commit - it enables the translations to be visible for some languages (es, fr, de) - which I think we want only when sufficient part will be translated already.

tokideveloper commented:
> You are correct. I want to get it working again first. It's almost there. For example [QubesOS/qubes-translated@a102a9e](https://github.com/QubesOS/qubes-translated/commit/a102a9e3cc6bfe7a1a73c647a18c0406e97c0616) was just made. There is still some issue with language switcher (just "en" visible there).

Currently (if language switcher enabled at all), for a certain page P, the language switcher only shows those languages where translations exist for P in `qubes-translated`. Should it be different?

marmarek commented:
> Currently (if language switcher enabled at all), for a certain page P, the language switcher only shows those languages where translations exist for P in `qubes-translated`.

I think this design is fine (even if may look weird at places). The issue is I see only "en" on pages I manually verified the "translated" version exists in the `qubes-translated` repo. For example documentation index, or `developer/building/qubes-builder.md`.

tokideveloper commented:
> > Currently (if language switcher enabled at all), for a certain page P, the language switcher only shows those languages where translations exist for P in `qubes-translated`.
> 
> I think this design is fine (even if may look weird at places). The issue is I see only "en" on pages I manually verified the "translated" version exists in the `qubes-translated` repo. For example documentation index, or `developer/building/qubes-builder.md`.

That's strange because I tested it on my machine a few minutes ago and it works. Okay, actually I only tested two files: `_doc/doc.md` and `_doc/user/how-to-guides/how-to-copy-and-paste-text.md`. And actually, I just copied the two original files from `qubesos.github.io` to the newly created `_qubes-translated` directory (with underscore), set `lang` and `permalink` appropriately and translated `title` test-wise. That was sufficient to work: On these two pages, the switcher contains both `en` and `de`, on the other pages not.

marmarek commented:
testdeploy

marmarek commented:
Ok, that was not very smart from me. "testdeploy" command took the version without downloaded translations :facepalm: Now it is better :)

tokideveloper commented:
> Now it is better :)

It's a lot better! It looks great! :tada: :heart:

marmarek commented:
https://wwwpreview.qubes-os.org/de/team/#inquiries - this doesn't look properly handled, especially "Business" should be "Geschäft" (it is translated in the qubes-translated repo). This is because `_includes/team.html` has:
```
  {% for team in site.data.team %}
    {% if team.type == "inquiries" %}
...
```

I can easily change that to appropriate entry from `site.translated`, but then `team.type` may not match "inquiries" if it gets translated. And in fact `team.type` `core` is translated to `Kern` in German version. I think `team.type` entry should _not_ be translated, as it's just an internal metadata for the layout - the actual section titles are in `team-page.yml` file.

Some ideas:
 - mark `team.type` entry as non translatable (locked?)
 - don't translate `team.yml` file at all, but move translatable strings into `team-page.yml` ("name" in the inquiries section, "role" in others)

I'm not sure how to do either of those.

marmarek commented:
Anyway, now is the time to review both the content of this very PR, but more importantly its output at wwwpreview.

As for the scripts in `_utils/_translation_utils`, some extra review would be useful, but not strictly necessary - those were mostly written by @maiska and @tokideveloper and already reviewed (and sometimes modified) by me.

tokideveloper commented:
> https://wwwpreview.qubes-os.org/de/team/#inquiries - this doesn't look properly handled, especially "Business" should be "Geschäft" (it is translated in the qubes-translated repo). This is because `_includes/team.html` has:
> 
> ```
>   {% for team in site.data.team %}
>     {% if team.type == "inquiries" %}
> ...
> ```
> 
> I can easily change that to appropriate entry from `site.translated`, but then `team.type` may not match "inquiries" if it gets translated. And in fact `team.type` `core` is translated to `Kern` in German version. I think `team.type` entry should _not_ be translated, as it's just an internal metadata for the layout - the actual section titles are in `team-page.yml` file.
> 
> Some ideas:
> 
> * mark `team.type` entry as non translatable (locked?)

Yes! Maybe other entries need to be locked, too.

> * don't translate `team.yml` file at all, but move translatable strings into `team-page.yml` ("name" in the inquiries section, "role" in others)

No. I like the current design where the team itself has its own data file.

> I'm not sure how to do either of those.

Me neither. @maiska, can you describe how to lock certain strings? Maybe via the API?

tokideveloper commented:
> Anyway, now is the time to review both the content of this very PR, but more importantly its output at wwwpreview.

I agree. First, the output should be good. Afterwards, we can review the code.

> As for the scripts in `_utils/_translation_utils`, some extra review would be useful, but not strictly necessary - those were mostly written by @maiska and @tokideveloper and already reviewed (and sometimes modified) by me.

In the past, I reviewed maiska's files and they seemed to be good. But I can do another review again.

tokideveloper commented:
> > * mark `team.type` entry as non translatable (locked?)
> 
> Yes! Maybe other entries need to be locked, too.

The smart tag "locked" means "nobody can touch it" and the smart tag "notranslate" means *in our case* "replace with source string". We should use both. See https://docs.transifex.com/projects/preventing-resource-edits#locking-strings

> Me neither. @maiska, can you describe how to lock certain strings? Maybe via the API?

See [_utils/_translation_utils/tag_strings_as_locked.py](https://github.com/QubesOS/qubesos.github.io/pull/154/files#diff-bb91fe7a7949258978696cdc6452107f31ff1baa3f1cdcd3690e3774cebd1a3b). You may need to load the diff manually.

tokideveloper reviewed (commented):
I've reviewed all files. At least, I tried it and tried to understand the files.

marmarek commented on _utils/_translation_utils/COUNTER.txt:
```diff
@@ -0,0 +1 @@
+current counter: 251
```
Yes, I'll rerun the script shortly after merging again. And generally, once it lands in master, the script will be called frequently.

marmarek commented on _config.yml:
```diff
@@ -21,6 +21,14 @@ relative_permalinks: false
+- de
+- fr
+- es
+
```
That's the plan. Adding those extra languages is a part of the "[DO NOT MERGE] enable languages: de fr es" commit.

marmarek commented on _utils/_translation_utils/merge_md_heading_ids.py:
```diff
@@ -0,0 +1,213 @@
+            result = merge_ids_in_gfm_files(original_lines, translated_lines)
+            write_lines(result, item)
+
+
```
Ah, right. I'll remove it.

marmarek commented on _utils/_translation_utils/merge_md_heading_ids.rb:
```diff
@@ -0,0 +1,335 @@
+        # headline starting with '#'
+        gfm_lines[line_number] = '# ' + placeholder + "\n"
+        hl = look_for_headline(render(gfm_lines), placeholder)
+    elsif next_line.length >= 3 and (line_only_made_of(next_line, '=') or line_only_made_of(next_line, '-'))
```
Good point.

marmarek commented on _utils/_translation_utils/merge_md_heading_ids.rb:
```diff
@@ -0,0 +1,335 @@
+
+    mapping = create_dict_from_tx_config(ARGV[0], ARGV[1])
+    mapping.each do |key, value|
+        if !key.end_with?(".yml")
```
Good point.

tokideveloper commented:
What I forget in the review: Some scripts run recursively through directories like `_doc` but don't ignore `.git`. To be corrected.

tokideveloper commented on _utils/_translation_utils/postprocess_htmlproofer.py:
```diff
@@ -0,0 +1,306 @@
+    mapping = {}
+    perms = []
+    yml_files = []
+    for dirname, subdirlist, filelist in walk(translated_dir):
```
This walk doesn't ignore `.git`.

tokideveloper commented on _utils/_translation_utils/postprocess_translation.py:
```diff
@@ -0,0 +1,450 @@
+
+    href = set()
+    reg ='(?<=href=\").*?(?=\")'
+    for dirname, subdirlist, filelist in walk(translated_dir):
```
This walk doesn't ignore `.git`.

tokideveloper commented on _utils/_translation_utils/postprocess_translation.py:
```diff
@@ -0,0 +1,450 @@
+    """
+
+    perms = set()
+    for dirname, subdirlist, filelist in walk(translated_dir):
```
This walk doesn't ignore `.git`.

tokideveloper commented on _utils/_translation_utils/prepare_for_translation.py:
```diff
@@ -0,0 +1,128 @@
+
+        return counter
+
+    for dir_name, subdir_list, file_list in os.walk(root_dir):
```
This walk doesn't ignore `.git`.

tokideveloper reviewed (commented):
In this review, I marked the spots where `.git` is not ignored.

Sometimes there seems to be no need to ignore `.git` since the root directory of such a walk is a repo subdirectory (not containing any `.git` directory). However, to be on the safe side, `.git` directories should be ignored in general.

jermanuts commented:
Any progress?

tokideveloper commented:
> Any progress?

This PR focuses on internationalization of the Qubes OS website including the documentation, using Markdown and Jekyll.

Meanwhile, we moved away from that approach towards a conversion of the documentation from Markdown to reStructuredText, render it using Sphinx and host it on ReadTheDocs. This is our (new) next step.

See https://github.com/QubesOS/qubes-doc/pull/1367

The next steps afterwards would be internationalization and localization of the documentation and later the Qubes OS website.
